File: scala/src/main/scala/leetcode/editor/cn/[567]字符串的排列.scala
package leetcode.editor.cn
object PermutationInString {
//leetcode submit region begin(Prohibit modification and deletion)
  object Solution {
    import scala.collection.mutable
    import scala.util.boundary
    def checkInclusion(s1: String, s2: String): Boolean = {
      val need = s1.groupBy(identity).view.mapValues(_.length).toMap
      val window = mutable.Map.empty[Char, Int]
      var left = 0
      var right = 0
      var valid = 0

      boundary {
        while (right < s2.length) {
          val c = s2(right)
          right += 1

          if (need.contains(c)) {
            window.updateWith(c) {
              case Some(v) =>
                Some(v+1)
              case None =>
                Some(1)
            }
            if (window(c) == need(c)) valid += 1
          }

          while (right - left >= s1.length) {
            if (valid == need.size)
              boundary.break(true)

            val d = s2(left)
            left += 1

            if (need.contains(d)) {
              window.updateWith(d) {
                case Some(v) =>
                  if (v == need(d)) valid -= 1
                  Some(v - 1)
                case None => None
              }
            }
          }
        }
        false
      }
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}

@main
def run(): Unit = {
  println(PermutationInString.Solution.checkInclusion("ab", "eidbaooo"))
}

File: cpp/leetcode/editor/cn/[567]字符串的排列.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <unordered_map>
#include <utility>
#include <memory>
#include <iterator>
using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    bool checkInclusion(string t, string s) {
        unordered_map<char,int> need;
        for (auto c : t) need[c]++;

        unordered_map<char,int> window;
        int valid = 0;
        int left = 0;
        int right = 0;

        while (right < s.size()){
            auto c = s[right];
            ++right;

            if (need.count(c)) {
                window[c]++;
                if (window[c] == need[c]) ++valid;
            }

            while (right - left >= t.size()){
                if (valid == need.size())
                    return true;

                auto d = s[left];
                ++left;

                if (need.count(d)) {
                    if (window[d] == need[d]) --valid;
                    window[d]--;
                }
            }
        }

        return false;

    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv,char *argc[]){
    Solution s;
    s.checkInclusion("ab","eidboaoo");
}